Hudl.Mjolnir/Command/CommandInvoker.cs

```diff
@@ -0,0 +1,639 @@
+
+                return CommandCompletionStatus.Canceled;
+            }
+
+            if (exception is CircuitBreakerRejectedException || exception is BulkheadRejectedException)
+            {
+                return CommandCompletionStatus.Rejected;
+            }
+
+            return CommandCompletionStatus.Faulted;
+        }
+
+        private static void AttachCommandExceptionData(BaseCommand command, Exception exception, CommandCompletionStatus status, InformativeCancellationToken ct)
+        {
+            // Use namespaced keys here to avoid clobbering data that the application might've
+            // already attached (or will attach later).
+            exception.WithData(new
+            {
+                MjolnirCommand = command.Name,
+                MjolnirStatus = status,
+                MjolnirBreaker = command.BreakerKey,
+                MjolnirBulkhead = command.BulkheadKey,
+                MjolnirTimeoutMillis = ct.DescriptionForLog,
+                MjolnirExecuteMillis = command.ExecutionTimeMillis,
+            });
+        }
+
+        private static bool IsCancellationException(Exception e)
+        {
+            return (e is TaskCanceledException || e is OperationCanceledException);
+        }
+
+        // "Failure" means any of [Fault || Timeout || Reject]
+        private enum OnFailure
+        {
+            Throw,
+            Return,
+        }
+    }
+
+    /// <summary>
+    /// Mostly for diagnostic purposes and logging, and should only be used internally. It's often
+    /// helpful to know the source of cancellation; when throwing an exception, we can be specific
+    /// in the message (e.g. "timed out" vs. just "canceled", which is clearer to the caller).
+    /// </summary>
+    internal struct InformativeCancellationToken
+    {
+        // Cancellation precedence
+        // - Config toggle to disable timeouts/cancellation
+        // - Cancellation token or timeout passed to Invoke()
+        // - Configured timeout for specific Command
+        // - Default timeout in Command constructor
+
+        private readonly CancellationToken _token;
+        private readonly TimeSpan? _timeout;
+        private readonly bool _isIgnored;
+
+        public CancellationToken Token { get { return _token; } }
+        public TimeSpan? Timeout { get { return _timeout; } }
+        public bool IsIgnored {  get { return _isIgnored; } }
+        
+        private InformativeCancellationToken(CancellationToken token, bool ignored = false)
+        {
+            _timeout = null;
+            _isIgnored = ignored;
+            _token = token;
+        }
+
+        private InformativeCancellationToken(TimeSpan timeout)
+        {
+            _timeout = timeout;
+            _isIgnored = false;
+
+            // This (int) cast probably breaks timeouts that are < 1 millisecond. I'm not sure
+            // that's worth fixing yet. Notable, though.
+            if ((int) timeout.TotalMilliseconds == 0)
```

Comment:
I think [it does fractional milliseconds](https://msdn.microsoft.com/en-us/library/system.timespan.totalmilliseconds(v=vs.110).aspx) as well:

> Gets the value of the current TimeSpan structure expressed in whole and fractional milliseconds.

Not sure if the actual presence of fractional milliseconds depends on the platform or OS, though.